### ISSUE TITLE
`mold` Start a clean compilation error!

### ISSUE THREAD

**Issue by zxlhhyccc** [openwrt/openwrt#15168]
### Describe the bug

````
mkdir -p /home/lin/ax6-6.1/staging_dir/host/stamp
touch /home/lin/ax6-6.1/build_dir/host/firmware-utils-2024.03.23~6b242991/.built
touch /home/lin/ax6-6.1/staging_dir/host/stamp/.firmware-utils_installed 
make[3]: Leaving directory '/home/lin/ax6-6.1/tools/firmware-utils'
time: tools/firmware-utils/compile#18.74#4.88#9.60
[426/479] Building C object third-party...ic.dir/__/__/__/lib/legacy/zstd_v01.c.o
In function 'ZSTDv01_decodeLiteralsBlock',
    inlined from 'ZSTD_decompressBlock' at /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1898:17:
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1504:20: warning: 'litbp.origSize' may be used uninitialized [-Wmaybe-uninitialized]
 1504 |             size_t rleSize = litbp.origSize;
      |                    ^~~~~~~
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c: In function 'ZSTD_decompressBlock':
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1488:23: note: 'litbp.origSize' was declared here
 1488 |     blockProperties_t litbp;
      |                       ^~~~~
In function 'ZSTDv01_decodeLiteralsBlock',
    inlined from 'ZSTD_decompressBlock' at /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1898:17:
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1495:5: warning: '*(unsigned int *)((char *)&litbp + offsetof(blockProperties_t, blockType))' may be used uninitialized [-Wmaybe-uninitialized]
 1495 |     switch(litbp.blockType)
      |     ^~~~~~
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c: In function 'ZSTD_decompressBlock':
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1488:23: note: '*(unsigned int *)((char *)&litbp + offsetof(blockProperties_t, blockType))' was declared here
 1488 |     blockProperties_t litbp;
      |                       ^~~~~
In function 'ZSTD_decompressSequences',
    inlined from 'ZSTD_decompressBlock' at /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1903:12:
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1849:35: warning: 'dumpsLength' may be used uninitialized [-Wmaybe-uninitialized]
 1849 |         seqState.dumpsEnd = dumps + dumpsLength;
      |                             ~~~~~~^~~~~~~~~~~~~
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c: In function 'ZSTD_decompressBlock':
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1825:23: note: 'dumpsLength' was declared here
 1825 |     size_t errorCode, dumpsLength;
      |                       ^~~~~~~~~~~
In function 'ZSTD_decompressSequences',
    inlined from 'ZSTD_decompressBlock' at /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1903:12:
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1849:35: warning: 'dumps' may be used uninitialized [-Wmaybe-uninitialized]
 1849 |         seqState.dumpsEnd = dumps + dumpsLength;
      |                             ~~~~~~^~~~~~~~~~~~~
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c: In function 'ZSTD_decompressBlock':
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1829:17: note: 'dumps' was declared here
 1829 |     const BYTE* dumps;
      |                 ^~~~~
In function 'ZSTD_decompressSequences',
    inlined from 'ZSTD_decompressBlock' at /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1903:12:
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1857:83: warning: 'nbSeq' may be used uninitialized [-Wmaybe-uninitialized]
 1857 |         for ( ; (FSE_reloadDStream(&(seqState.DStream)) <= FSE_DStream_completed) && (nbSeq>0) ; )
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c: In function 'ZSTD_decompressBlock':
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1828:9: note: 'nbSeq' was declared here
 1828 |     int nbSeq;
      |         ^~~~~
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c: In function 'ZSTDv01_decompressDCtx':
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1935:9: warning: '*(unsigned int *)((char *)&blockProperties + offsetof(blockProperties_t, blockType))' may be used uninitialized [-Wmaybe-uninitialized]
 1935 |         switch(blockProperties.blockType)
      |         ^~~~~~
/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/zstd/lib/legacy/zstd_v01.c:1917:23: note: '*(unsigned int *)((char *)&blockProperties + offsetof(blockProperties_t, blockType))' was declared here
 1917 |     blockProperties_t blockProperties;
      |                       ^~~~~~~~~~~~~~~
[453/479] Building CXX object third-par.../src/tbb/CMakeFiles/tbb.dir/arena.cpp.o
FAILED: third-party/tbb/src/tbb/CMakeFiles/tbb.dir/arena.cpp.o 
/home/lin/ax6-6.1/staging_dir/host/bin/g++ -D__TBB_BUILD -D__TBB_DYNAMIC_LOAD_ENABLED=0 -D__TBB_SOURCE_DIRECTLY_INCLUDED=1 -D__TBB_USE_ITT_NOTIFY -I/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/tbb/src/tbb/../../include -O2 -I/home/lin/ax6-6.1/staging_dir/host/include -DNDEBUG -std=c++11 -fPIC -flifetime-dse=1 -Wall -Wextra -Wfatal-errors -mrtm -mwaitpkg -fstack-clash-protection -fno-strict-overflow -fno-delete-null-pointer-checks -fwrapv -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -D_FORTIFY_SOURCE=2 -pthread -MD -MT third-party/tbb/src/tbb/CMakeFiles/tbb.dir/arena.cpp.o -MF third-party/tbb/src/tbb/CMakeFiles/tbb.dir/arena.cpp.o.d -o third-party/tbb/src/tbb/CMakeFiles/tbb.dir/arena.cpp.o -c /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/tbb/src/tbb/arena.cpp
In file included from /usr/include/c++/13/atomic:41,
                 from /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/tbb/src/tbb/../../include/oneapi/tbb/detail/_utils.h:22,
                 from /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/tbb/src/tbb/task_dispatcher.h:20,
                 from /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/tbb/src/tbb/arena.cpp:17:
In member function 'void std::__atomic_base<_IntTp>::store(__int_type, std::memory_order) [with _ITp = bool]',
    inlined from 'void std::atomic<bool>::store(bool, std::memory_order)' at /usr/include/c++/13/atomic:104:20,
    inlined from 'void tbb::detail::r1::concurrent_monitor_base<Context>::notify_one_relaxed() [with Context = long unsigned int]' at /home/lin/ax6-6.1/build_dir/host/mold-2.3.2/third-party/tbb/src/tbb/concurrent_monitor.h:293:53:
/usr/include/c++/13/bits/atomic_base.h:481:25: warning: 'void __atomic_store_1(volatile void*, unsigned char, int)' writing 1 byte into a region of size 0 overflows the destination [-Wstringop-overflow=]
  481 |         __atomic_store_n(&_M_i, __i, int(__m));
      |         ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
In member function 'void tbb::detail::r1::concurrent_monitor_base<Context>::notify_one_relaxed() [with Context = long unsigned int]':
cc1plus: note: destination object is likely at address zero
/home/lin/ax6-6.1/tmp/ccrFmFkY.s: Assembler messages:
/home/lin/ax6-6.1/tmp/ccrFmFkY.s:8512: Error: no such instruction: `tpause %r13d'
/home/lin/ax6-6.1/tmp/ccrFmFkY.s:8977: Error: no such instruction: `tpause %r13d'
[456/479] Building CXX object third-par.../tbb/CMakeFiles/tbb.dir/exception.cpp.o
ninja: build stopped: subcommand failed.
Makefile:22: recipe for target '/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/.built' failed
make[3]: *** [/home/lin/ax6-6.1/build_dir/host/mold-2.3.2/.built] Error 1
make[3]: Leaving directory '/home/lin/ax6-6.1/tools/mold'
time: tools/mold/compile#1028.97#119.99#299.24
    ERROR: tools/mold failed to build.
tools/Makefile:231: recipe for target 'tools/mold/compile' failed
make[2]: *** [tools/mold/compile] Error 1
make[2]: Leaving directory '/home/lin/ax6-6.1'
tools/Makefile:229: recipe for target '/home/lin/ax6-6.1/staging_dir/host/stamp/.tools_compile_nyyynyynnnyyynynyyyyyyynyyynnynyynnynnyyyyyyyyyyyyyyyyyynynnyyyyyyyyyy' failed
make[1]: *** [/home/lin/ax6-6.1/staging_dir/host/stamp/.tools_compile_nyyynyynnnyyynynyyyyyyynyyynnynyynnynnyyyyyyyyyyyyyyyyyynynnyyyyyyyyyy] Error 2
make[1]: Leaving directory '/home/lin/ax6-6.1'
Build failed - please re-run with -j1 to see the real error message
/home/lin/ax6-6.1/include/toplevel.mk:231: recipe for target 'world' failed
make: *** [world] Error 1
````

### OpenWrt version

r25871-d668c74fe6

### OpenWrt release

r25871-d668c74fe6

### OpenWrt target/subtarget

ipq807x/generic

### Device

Redmi AX6

### Image kind

Self-built image

### Terms

- [X] I am reporting an issue for OpenWrt, not an unsupported fork.

**Comment by robimarko** [comment]
Provide diffconfig as I just cannot reproduce this

**Comment by robimarko** [comment]
Since you provided that you are using 18.04 with GCC13.2 but with binutils 2.30 that is your issue again.

Either update your binutils or your distribution as 18.04 is not supported.

**Comment by robimarko** [comment]
@zxlhhyccc Any luck with newer binutils?

**Comment by zxlhhyccc** [comment]
@robimarko The current latest version of 18.04 is 2.32. Can the latest version be installed through compilation?

**Comment by robimarko** [comment]
@zxlhhyccc Yes, I just wasted some time building libdeflate under 18.04 with GCC13.1 from the test toolchain PPA.

It will fail due to outdated binutils, but then I built binutils 2.36 locally with `--prefix=/usr/local` and installed them and then it compiles fine.

So, there is no way around this other than updating your binutils since it was just a matter of time to hit the outdated binutils being an issue.
